src/software/thunderscope/robot_diagnostics/robot_error_log.py

```diff
@@ -0,0 +1,183 @@
+                file_name="file.txt",
+                log_msg="test",
+            )
+        )
+
+    def refresh(self):
+        """
+        Refreshes the widget's graphics
+
+        Gets the RobotCrash and RobotStatus messages from the buffers and adds
+        the corresponding log messages if needed
+
+        Updates the time since of the existing messages
+        """
+        # update time since for all existing logs
+        for error_message in self.error_log_messages:
+            error_message.update_time()
+
+        # get the robot log message
+        robot_log = self.robot_log_buffer.get(block=False, return_cached=False)
+
+        if robot_log is not None:
+            if (
+                robot_log.log_level == LogLevel.FATAL
+                or robot_log.log_level == LogLevel.CONTRACT
+            ):
+                if (
+                    robot_log.robot_id not in self.robot_last_fatal_time_s
+                    or time.time() - self.robot_last_fatal_time_s[robot_log.robot_id]
+                    > ROBOT_FATAL_TIMEOUT_S
+                ):
+                    self.add_error_log_message(FatalLogMessageWidget(robot_log))
+                self.robot_last_fatal_time_s[robot_log.robot_id] = time.time()
+
+        # get the robot crash message
+        robot_crash = self.robot_crash_buffer.get(block=False, return_cached=False)
+
+        # if there's a crash, log it
+        if robot_crash is not None:
+            if (
+                robot_crash.robot_id not in self.robot_last_crash_time_s
+                or time.time() - self.robot_last_crash_time_s[robot_crash.robot_id]
+                > ROBOT_CRASH_TIMEOUT_S
+            ):
+                self.add_error_log_message(RobotCrashLogMessageWidget(robot_crash))
+            # prevent repeated crash logs by having a buffer time
+            self.robot_last_crash_time_s[robot_crash.robot_id] = time.time()
+
+        # get a RobotStatus message
+        robot_status = self.robot_status_buffer.get(block=False, return_cached=False)
+
+        # empty out the buffer to get statuses for all current robots at once
+        while robot_status is not None:
+            # if there's an error code, add to log
+            for code in robot_status.error_code:
+                if (
+                    code != ErrorCode.NO_ERROR
+                    and code != ErrorCode.LOW_BATTERY
+                    and code in ERROR_CODE_MESSAGES.keys()
+                ):
+                    self.add_error_log_message(
+                        ErrorCodeLogMessageWidget(
+                            robot_status.robot_id, ERROR_CODE_MESSAGES[code]
+                        )
+                    )
+
+            # if the battery voltage is too low, log it
+            if (
+                robot_status.power_status.battery_voltage <= BATTERY_WARNING_VOLTAGE
+                and not self.low_battery_log_disabled
+            ):
+                self.add_error_log_message(
```

Comment:
Maybe I am understanding this logic wrong but would this mean that if two robots had low batteries, then only one error log would show up?

Reply:
good catch, fixed it now